tests/rust/aligned_struct.rs

```diff
@@ -0,0 +1,84 @@
+#[repr(packed, C)]
+pub struct PackedStruct {
+    pub arg1: usize,
+    pub arg2: *mut u8,
+}
+
+#[repr(align(1), C)]
+pub struct Align1Struct {
+    pub arg1: usize,
+    pub arg2: *mut u8,
+}
+
+#[repr(align(2), C)]
+pub struct Align2Struct {
+    pub arg1: usize,
+    pub arg2: *mut u8,
+}
+
+#[repr(align(4), C)]
+pub struct Align4Struct {
+    pub arg1: usize,
+    pub arg2: *mut u8,
+}
+
+#[repr(align(8), C)]
+pub struct Align8Struct {
+    pub arg1: usize,
+    pub arg2: *mut u8,
+}
+
+#[repr(align(32), C)]
+pub struct Align32Struct {
+    pub arg1: usize,
+    pub arg2: *mut u8,
+}
+
+// Packed doesn't take an argument.
+#[repr(packed(4), C)]
```

Comment:
Actually it can, see https://doc.rust-lang.org/reference/type-layout.html.

We should either `warn!` or add a third `packed_n` annotation.

Reply:
Yeah, I can implement `packed_n` as well.

Reply:
Feel free to do it as a separate PR if more convenient, that way it's easier to review too :)

Reply:
It turns out this is slightly more complicated to do correctly, so I will punt on this for now and do it in a follow-up PR.

If I'm reading [this](https://doc.rust-lang.org/nightly/reference/type-layout.html#the-alignment-modifiers) right, it turns out that `#[repr(packed(n))]` does not really have a simple native C representation. In Rust, `#[repr(packed(n))]` indicates that the fields should have the smaller of `n` and `std::mem::align_of<field_type>()`. While in C you can specify individual field alignments like so:

```
struct foo {
  uint64_t foo1 __attribute__((aligned(32)));
  uint32_t foo2 __attribute__((aligned(32)));
  uint16_t foo3 __attribute__((aligned(32)));
  uint8_t foo4 __attribute__((aligned(32)));
  bool foo5 __attribute__((aligned(32)));
} __attribute__((packed));
```

The issue is that, in Rust, the following struct:

```
#[repr(C, packed(32))]
struct Foo {
  foo1: u64;
  foo2: u32;
  foo3: u16;
  foo4: u8;
  foo5: bool;
}
```

Should actually be represented as (depending on your architecture):

```
struct foo {
  uint64_t foo1 __attribute__((aligned(8)));
  uint32_t foo2 __attribute__((aligned(4)));
  uint16_t foo3 __attribute__((aligned(2)));
  uint8_t foo4 __attribute__((aligned(1)));
  bool foo5 __attribute__((aligned(1)));
} __attribute__((packed));
```

And the issue is that it's not clear to me how we can determine what the natural alignment of the constituent types is. And unfortunately, this is critical to making sure that C and Rust code can interoperate with packed structures. It's a bit unfortunate that Rust decided to have different semantics to GCC's alignment extensions, because I'm not really sure how to implement them. And given that `std::mem::align_of::<>()` depends on your architecture, we'd have to do lots of extra work to generate the right alignment for each architecture rust supports.

Reply:
Filed the above as #433.

Reply:
Thanks for filing that! Yeah, I wonder how far can you go with `alignas` and decltype, if you're using C++...